flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/aggregate/RowTimeBoundedRangeOver.java

```diff
@@ -0,0 +1,279 @@
+		this.accTypes = accTypes;
+		this.inputFieldTypes = inputFieldTypes;
+		this.precedingOffset = precedingOffset;
+		this.rowTimeIdx = rowTimeIdx;
+	}
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		LOG.debug("Compiling AggregateHelper: " + genAggsHandler.getClassName() + " \n\n" +
+			"Code:\n" + genAggsHandler.getCode());
+		function = genAggsHandler.newInstance(getRuntimeContext().getUserCodeClassLoader());
+		function.open(new PerKeyStateDataViewStore(getRuntimeContext()));
+
+		output = new JoinedRow();
+
+		ValueStateDescriptor<Long> lastTriggeringTsDescriptor = new ValueStateDescriptor<Long>(
+			"lastTriggeringTsState",
+			Types.LONG);
+		lastTriggeringTsState = getRuntimeContext().getState(lastTriggeringTsDescriptor);
+
+		BaseRowTypeInfo accTypeInfo = new BaseRowTypeInfo(accTypes);
+		ValueStateDescriptor<BaseRow> accStateDesc = new ValueStateDescriptor<BaseRow>("accState", accTypeInfo);
+		accState = getRuntimeContext().getState(accStateDesc);
+
+		// input element are all binary row as they are came from network
+		BaseRowTypeInfo inputType = new BaseRowTypeInfo(inputFieldTypes);
+		ListTypeInfo<BaseRow> rowListTypeInfo = new ListTypeInfo<BaseRow>(inputType);
+		MapStateDescriptor<Long, List<BaseRow>> inputStateDesc = new MapStateDescriptor<Long, List<BaseRow>>(
+			"inputState",
+			Types.LONG,
+			rowListTypeInfo);
+		inputState = getRuntimeContext().getMapState(inputStateDesc);
+
+		initCleanupTimeState("RowTimeBoundedRangeOverCleanupTime");
+	}
+
+	@Override
+	public void processElement(
+		BaseRow input,
+		KeyedProcessFunction<K, BaseRow, BaseRow>.Context ctx,
+		Collector<BaseRow> out) throws Exception {
+		// register state-cleanup timer
+		registerProcessingCleanupTimer(ctx, ctx.timerService().currentProcessingTime());
+
+		// triggering timestamp for trigger calculation
+		Long triggeringTs = input.getLong(rowTimeIdx);
+
+		Long lastTriggeringTs = lastTriggeringTsState.value();
+		if (lastTriggeringTs == null) {
+			lastTriggeringTs = 0L;
+		}
+
+		// check if the data is expired, if not, save the data and register event time timer
+		if (triggeringTs > lastTriggeringTs) {
+			List<BaseRow> data;
+			data = inputState.get(triggeringTs);
+			if (null != data) {
+				data.add(input);
+				inputState.put(triggeringTs, data);
+			} else {
+				data = new ArrayList<BaseRow>();
+				data.add(input);
+				inputState.put(triggeringTs, data);
+				// register event time timer
+				ctx.timerService().registerEventTimeTimer(triggeringTs);
+			}
+		}
+	}
+
+	@Override
+	public void onTimer(
+		long timestamp,
+		KeyedProcessFunction<K, BaseRow, BaseRow>.OnTimerContext ctx,
+		Collector<BaseRow> out) throws Exception {
+		// register state-cleanup timer
+		registerProcessingCleanupTimer(ctx, ctx.timerService().currentProcessingTime());
+
+		if (isProcessingTimeTimer(ctx)) {
+			if (needToCleanupState(timestamp)) {
+
+				Iterator<Map.Entry<Long, List<BaseRow>>> keysIt = inputState.iterator();
```

Comment:
Use `inputState.keys()` instead of `inputState.iterator()` to avoid unnecessary cost.